### ISSUE TITLE
Support for responsive props

### ISSUE THREAD

**Issue by afzalsayed96** [callstack/linaria#708]
## Describe the feature

Is there a way to implement responsive props with linaria similar to https://github.com/emotion-js/facepaint?

## Motivation

When implemening a primitive component using `Box` or `Text`, it is absolutely necessary to support responsive behaviour.
Currently it is only possible to control props supplied to component externally such as

```js
const isMobile = useMediaQuery(breakpoints.MOBILE)

<Box padding={isMobile ? 20 : 40} />
```

## Possible implementations

It would be great if we could do something like:

```js
<Box padding={[20, 40]} />
```

which could generate css media queries based on a static config.

## Related Issues

Probably #465

**Comment by MaxSvargal-imhio** [comment]
Like in Theme-UI. I like it!

**Comment by mwojslaw** [comment]
Is that something linaria has on its roadmap? I'm struggling with the same use case as mentioned by @afzalsayed96, I think this pattern is quite common, and really helpful when building kit libraries. Look at mentioned Theme-UI, styled-system, rebassjs, chakra-ui, and probably much more.

**Comment by Anber** [comment]
Idk how it can be implemented without runtime and without multiplying the size of generated styles. I mean, we can probably do something like this:
```js
const cssWithMq = facepaint([
  '@media(min-width: 420px)',
  '@media(min-width: 920px)',
  '@media(min-width: 1120px)'
])

const myClassName = cssWithMq`
  background-color: 'hotpink';
  text-align: 'center';
  width: ${['25%', '50%', '75%', '100%']};
  & .foo: {
    color: ${['red', 'green', 'blue', 'darkorchid']};
    & img: {
      height: ${[10, 15, 20, 25]};
    }
  }
`;
```

but since we don't parse css, all we can do is repeat that block 4 times with media queries. And that will quadruple the size of generated css. It's not a problem for styled-components and emotion, because they generate styles on a client-side and in that case size doesn't really matter.

**Comment by mwojslaw** [comment]
Yup, probably need to wait until we could use css properties in media queries like this:

`@media (min-width: --theme-breakpoint-sm}) {
        color: blue;
}`

**Comment by Anber** [comment]
Probably, we can generate something like this:
```
@media(min-width: 420px) {
  --some-generated-variable-for-color-in-myClassName: 'green';
}

@media(min-width: 920px) {
  --some-generated-variable-for-color-in-myClassName: 'blue';
}

@media(min-width: 1120px) {
  --some-generated-variable-for-color-in-myClassName: 'darkorchid';
}

. myClassName {
  color: var(--some-generated-variable-for-color-in-myClassName, 'red');
}
```

**Comment by mwojslaw** [comment]
Yes, but the problem is occurring when those breakpoints are customizable, not static. Usually passed as theme properties. I'm assuming it can't be done at compilation time and requires some runtime to work.

**Comment by Anber** [comment]
Yep, you are right. It will work only for constants.

**Comment by devinrhode2** [comment]
I wonder if there's a good pattern to integrate with an existing runtime-library for adding breakpoints into the render tree. 

https://www.npmjs.com/package/@artsy/fresnel#why-not-conditionally-render Is nice due to it's SSR support, but it's also a bit odd that it simply ships server-rendered html for all breakpoints. Almost works against the whole point of ssr... performance.

React-responsive has some nice hooks for media queries.

**Comment by Anber** [comment]
@devinrhode2 
I don't sure that it's what you are looking for but [that plugin](https://github.com/anber/linaria-interop) allows to use Linaria and styled/emotion at the same time.

**Comment by fabb** [comment]
> It's not a problem for styled-components and emotion, because they generate styles on a client-side and in that case size doesn't really matter.

Not true for SSR and SSG.

**Comment by Anber** [comment]
> Not true for SSR and SSG.
That's a good point. But do they really care about the size of styles in that case?

**Comment by engelmav** [comment]
You can kind of do this:

```
const breakpoints = [
  ["max-width", "40em"],
  ["min-width", "52em"],
  ["min-width", "64em"],
  ["min-width", "80em"]
]
```

```
const fontSteps = {
  1: "10px",
  2: "90px",
  3: "15px",
  4: "25px",
  6: "30px",
  7: "40px"
};
```
```
export function fontSize(values){
  const styles = {};
  breakpoints.forEach((breakpoint, idx) => {
    styles[`@media (${breakpoint[0]}: ${breakpoint[1]})`] = {
      fontSize: fontSteps[values[idx]],
    };
  });
  return styles;
}
```

```
export const H1 = styled.h1`
 ... other styles ...
  ${fontSize([4, 6, 7])}
`;
```

**Comment by joseDaKing** [comment]
@Anber can not this be achieved at the component level by using matchMedia under the hood

**Comment by Anber** [comment]
@Anber will it make Linaria a non-zero-runtime library? :)

**Comment by fabb** [comment]
> Idk how it can be implemented without runtime and without multiplying the size of generated styles. I mean, we can probably do something like this:
> 
> ```js
> const cssWithMq = facepaint([
>   '@media(min-width: 420px)',
>   '@media(min-width: 920px)',
>   '@media(min-width: 1120px)'
> ])
> 
> const myClassName = cssWithMq`
>   background-color: 'hotpink';
>   text-align: 'center';
>   width: ${['25%', '50%', '75%', '100%']};
>   & .foo: {
>     color: ${['red', 'green', 'blue', 'darkorchid']};
>     & img: {
>       height: ${[10, 15, 20, 25]};
>     }
>   }
> `;
> ```
> 
> but since we don't parse css, all we can do is repeat that block 4 times with media queries. And that will quadruple the size of generated css.

Would it be an option to start parsing the css string, or would compile times explode? Compile times with styled-components SSR is fine. With [object interpolations](https://github.com/callstack/linaria/blob/master/docs/BASICS.md#object-interpolations) it would even be simpler, no css parsing necessary but I guess we need a solution that works for template literals too.

maybe an alternative would be something like this? Not as nice to read though:
```
const medias = [
  '@media(min-width: 420px)',
  '@media(min-width: 920px)',
  '@media(min-width: 1120px)'
]

const responsive = (attribute, responsiveValues) => responsiveValues.map((value, index) => index === 0 ? `${attribute}: ${value}` : `${medias[index-1]} { ${attribute}: ${value} }`).join(';'))

const myClassName = css`
  background-color: 'hotpink';
  text-align: 'center';
  ${responsive('width', ['25%', '50%', '75%', '100%']};
`;
```

**Comment by devinrhode2** [comment]
I don't think it's a silver bullet, but it may be worth seeing how the 4x larger css compresses under gzip, and then serving that with an http2/spdy server.. and of course, browser parse time will also need to be measured

**Comment by espretto** [comment]
One thing to consider is CSS' [attr](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/attr) function which will extend support beyond CSS property `content` to be used almost anywhere. That could mean we'll be writing more CSS than JavaScript to map DOM attributes to styles in the future.
